persist/fs/read.go

```diff
@@ -63,43 +66,159 @@ type reader struct {
 	blockSize      time.Duration
 
 	infoFdWithDigest           digest.FdWithDigestReader
-	indexFdWithDigest          digest.FdWithDigestReader
-	dataFdWithDigest           digest.FdWithDigestReader
 	digestFdWithDigestContents digest.FdWithDigestContentsReader
-	expectedInfoDigest         uint32
-	expectedIndexDigest        uint32
-	expectedDataDigest         uint32
-	expectedDigestOfDigest     uint32
-
-	unreadBuf   []byte
-	entries     int
-	entriesRead int
-	prologue    []byte
-	decoder     encoding.Decoder
-	digestBuf   digest.Buffer
-	bytesPool   pool.CheckedBytesPool
+
+	indexFd     *os.File
+	indexMmap   []byte
+	indexReader *decoderStream
+
+	dataFd     *os.File
+	dataMmap   []byte
+	dataReader digest.ReaderWithDigest
+
+	expectedInfoDigest     uint32
+	expectedIndexDigest    uint32
+	expectedDataDigest     uint32
+	expectedDigestOfDigest uint32
+	entries                int
+	entriesRead            int
+	metadataRead           int
+	prologue               []byte
+	decoder                encoding.Decoder
+	digestBuf              digest.Buffer
+	bytesPool              pool.CheckedBytesPool
+
+	indexEntriesByOffset []schema.IndexEntry
+}
+
+type indexEntriesByOffset []schema.IndexEntry
+
+func (e indexEntriesByOffset) Len() int {
+	return len(e)
+}
+
+func (e indexEntriesByOffset) Less(i, j int) bool {
+	return e[i].Offset < e[j].Offset
+}
+
+func (e indexEntriesByOffset) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+type decoderStream struct {
+	bytesReader      *bytes.Reader
+	readerWithDigest digest.ReaderWithDigest
+	backingBytes     []byte
+	buf              [8]byte
+	lastReadByte     int
+	unreadByte       int
+}
+
+func newDecoderStream() *decoderStream {
```

Comment:
How come you're not reusing `encoding.NewDecoderStream` here?

Reply:
Need special logic to call the reader with digest.  The encoding `NewDecoderStream(...)` is for more simple use cases.

Reply:
afaict, most of the code except `Skip()` is the same. Any chance of re-using those bits?